tools/tmt/Loader.cs

```diff
@@ -0,0 +1,262 @@
+
+using Xamarin.Tools.Zip;
+
+namespace tmt
+{
+	class Loader
+	{
+		const string xamarinApp = "libxamarin-app.so";
+		const string sharedLibsDirName = "app_shared_libraries";
+
+		static readonly List<string> archDirs = new List<string> {
+			"arm64-v8a",
+			"armeabi-v7a",
+			"x86",
+			"x86_64",
+		};
+
+		static readonly char dirSep = Path.DirectorySeparatorChar;
+
+		AndroidArch archFilter;
+		bool loadOnlyFirst;
+
+		public Loader (AndroidArch archFilter, bool loadOnlyFirst)
+		{
+			this.archFilter = archFilter;
+			this.loadOnlyFirst = loadOnlyFirst;
+		}
+
+		public List<ITypemap> TryLoad (string path)
+		{
+			if (File.Exists (path)) {
+				return LoadFromFile (path);
+			}
+
+			if (!Directory.Exists (path)) {
+				Log.Info ($"Directory or file '{path}' not found.");
+				return new List<ITypemap> ();
+			}
+
+			string fsPath = Path.Combine (path, xamarinApp);
+			if (File.Exists (fsPath)) {
+				return LoadFromFile (fsPath);
+			}
+
+			if (PathEndsWith (path, sharedLibsDirName)) {
+				return LoadFromSharedLibsDir (path);
+			}
+
+			fsPath = Path.Combine (path, sharedLibsDirName);
+			if (Directory.Exists (fsPath)) {
+				return LoadFromSharedLibsDir (fsPath);
+			}
+
+			if (PathEndsWith (path, "obj")) {
+				return LoadFromObjDir (path);
+			}
+
+			fsPath = Path.Combine (path, "obj");
+			if (Directory.Exists (path)) {
+				return LoadFromObjDir (fsPath);
+			}
+
+			Log.Warning ($"Cannot determine what to load from {path}");
+			return new List<ITypemap> ();
+		}
+
+		List<ITypemap> LoadFromObjDir (string objDirPath)
+		{
+			var ret = new List<ITypemap> ();
+			foreach (string subdir in Directory.EnumerateDirectories (objDirPath)) {
+				string fsPath = Path.Combine (subdir, sharedLibsDirName);
+				if (!Directory.Exists (fsPath)) {
+					continue;
+				}
+
+				ret.AddRange (LoadFromSharedLibsDir (fsPath));
+			}
+
+			return ret;
+		}
+
+		List<ITypemap> LoadFromSharedLibsDir (string dirPath)
+		{
+			var ret = new List<ITypemap> ();
+			foreach (string archDir in archDirs) {
+				string fsPath = Path.Combine (dirPath, archDir, xamarinApp);
+				if (File.Exists (fsPath)) {
+					ITypemap? dsoTypemap = LoadDSOFromFilesystem (fsPath);
+					if (dsoTypemap != null) {
+						ret.Add (dsoTypemap);
+						if (loadOnlyFirst) {
+							break;
+						}
+					}
+				}
+			}
+
+			return ret;
+		}
+
+		List<ITypemap> LoadFromFile (string filePath)
+		{
+			string ext = Path.GetExtension (filePath);
+			if (String.Compare (ext, ".so", StringComparison.OrdinalIgnoreCase) == 0) {
+				ITypemap? dsoTypemap = LoadDSOFromFilesystem (filePath);
+				if (dsoTypemap == null) {
+					return new List<ITypemap> ();
+				}
+
+				return new List<ITypemap> {
+					dsoTypemap
+				};
+			}
+
+			if (String.Compare (ext, ".apk", StringComparison.OrdinalIgnoreCase) == 0) {
+				return LoadFromAPK (filePath);
+			}
+
+			if (String.Compare (ext, ".aab", StringComparison.OrdinalIgnoreCase) == 0) {
+				throw new NotImplementedException ("AAB");
```

Comment:
abb should be pretty similar to apk I think. I suspect the only difference is the path which has a `root` folder which then contains the `lib` folders.

Reply:
Yeah, I need to take a look at one :)